### PR TITLE
GroupBy Aggregate with StdDev function.

### PR DESCRIPTION
**Pull request by nitinmotgi**

### CODE REVIEW

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  private double sum1;
+  private double sum2;
+  private int count;
+  private double stddev;
+
+  public Stddev(String fieldName, Schema fieldSchema) {
+    this.fieldName = fieldName;
+    boolean isNullable = fieldSchema.isNullable();
+    Schema.Type fieldType = isNullable ? fieldSchema.getNonNullable().getType() : fieldSchema.getType();
+    if (!(fieldType == Schema.Type.INT || fieldType == Schema.Type.LONG ||
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: This looks like an often reused check, would it make sense to make it part of the type?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    sum2 += Math.pow(value, 2);
+    stddev = Math.sqrt(count*sum2 - Math.pow(sum1, 2));
+    count++;
+  }
+
+  @Override
+  public Double getAggregate() {
+    if (count == 0) {
+      // only happens if the field value was always null
+      return null;
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: This looks dangerous, should the interface method return an Optional instead?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  @Override
+  public void operateOn(StructuredRecord record) {
+    Object val = record.get(fieldName);
+    if (val == null) {
+      return;
+    }
+    double value = ((Number) val).doubleValue();
+    sum1 += value;
+    sum2 += Math.pow(value, 2);
+    stddev = Math.sqrt(count*sum2 - Math.pow(sum1, 2));
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: Why do we need to compute the `stddev` on each record?
Much more efficient to compute it only once (and cache it?) when it's used.

Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: This technique allows you to come compute stddev in one pass, otherwise you first have to calculate mean and then do the sum of squared. The interface doesn't allow you to come that way. So, had to find techniques that allow you to compute stddev in a single pass over the numbers.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 2 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: What you are saying makes sense, but, that would need an interface change. I am sure @albertshau might have thought of this. So, I am sticking to the interface as it is designed right now.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 1 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: Schema doesn't have comparators for specific types right. It should a utility or common function. But, you are right that it gets used multiple times in multiple places. I will move it into a common function.

**Comment by nitinmotgi** [comment]
@Nishith @anwar6953 please review again. Thanks.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+package co.cask.hydrator.plugin.batch.aggregator;
+
+import co.cask.cdap.api.data.schema.Schema;
+
+/**
+ * Common functions for aggregation related functionalities.
+ */
+public final class AggrCommon {
```
Comment (chtyim) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java]: Is this a common util class? If it is, call it `AggregationUtils` and add a private constructor.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+   * Checks if the type of the field is Numeric.
+   * @param fieldType type of the field to be checked.
+   * @return true if numeric, false otherwise.
+   */
+  public static boolean isNumericType(Schema.Type fieldType) {
+    if (fieldType == Schema.Type.INT || fieldType == Schema.Type.LONG ||
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java]: How about
public static boolean isNumericType(Schema.Type fieldType) {
    return (fieldType == Schema.Type.INT || fieldType == Schema.Type.LONG ||  fieldType == Schema.Type.FLOAT ||    fieldType == Schema.Type.DOUBLE);
}

You can keep what you have if you think that's more readable.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      fieldType == Schema.Type.FLOAT || fieldType == Schema.Type.DOUBLE) {
+      return true;
+    }
+    return false;
```
Comment (chtyim) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java]: You can simplify it to `return (field == Schema.Type.INT) || ....` instead of `if`.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * Calculates the Standard Deviation
+ */
+public class Stddev implements AggregateFunction<Double> {
+  private final String fieldName;
+  private final Schema outputSchema;
+  private double sum1;
+  private double sum2;
+  private int count;
```
Comment (chtyim) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: Should it be `long`?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.hydrator.plugin.batch.aggregator.AggrCommon;
+
+/**
+ * Calculates the Standard Deviation
+ */
+public class Stddev implements AggregateFunction<Double> {
+  private final String fieldName;
+  private final Schema outputSchema;
+  private double sum1;
```
Comment (chtyim) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: why the sum is double instead of long?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 3 above)
Comment (chtyim) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: how about this? http://www.johndcook.com/blog/standard_deviation/

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 2 above)
Comment (chtyim) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: The interface doesn't have `@Nullable`, so I don't think it's safe to return `null`. How about returning `0` instead?

Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: 0 Might be a valid value, perhaps Double.NaN?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 3 above)
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: In the current implementation (lines 59-62), it seems like `stddev` should be able to be computed just once at the end.
Just build up `sum1` and `sum2` as you are now (they don't depend on `stddev`, and neither does `stddev` depend on its previous values). And then at the end, perform the statement on line 62.
Am I missing something?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java (hunk 5 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java]: done!

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java (hunk 6 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java]: done!

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 7 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: done!

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 8 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: values could be real

**Comment by Nishith** [comment]
LGTM.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 2 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: Using Double.NaN

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java (hunk 3 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Stddev.java]: Going to use the new implementation that @chtyim suggested.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java (hunk 4 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggrCommon.java]: done!

**Comment by nitinmotgi** [comment]
@chtyim and @Nishith one last review please.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.aggregator.function;
+
+public final class RunningStats  {
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java:24]: javadoc probably needed, for checkstyle, right?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    }  else  {
+      newM = oldM + (x - oldM) / entries;
+      newS = oldS + (x - oldM) * (x - newM);
+      // set up for next iteration
+      oldM = newM;
+      oldS = newS;
+    }
+  }
+
+  public long values() {
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: this is more like `numValues` or `numEntries`.
Method name like `values` makes it seem like it returns the actual values.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Variance.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+  @Override
+  public void beginFunction() {
+    stats = new RunningStats();
+  }
+
+  @Override
+  public void operateOn(StructuredRecord record) {
+    Object val = record.get(fieldName);
+    if (val == null) {
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Variance.java:50]: Would a user ever want this to be equivalent to a `push(0)`?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java [hunk 12]
```diff
@@ -156,7 +156,7 @@
   }
 
   /**
-   * AggrCommon properties for BatchWritable source and sinks
+   * AggregationUtils properties for BatchWritable source and sinks
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java]: this javadoc seems wrong/confusing.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggregationUtils.java [hunk 13]
```diff
@@ -21,19 +21,18 @@
 /**
  * Common functions for aggregation related functionalities.
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggregationUtils.java]: The function(s) defined in this utility class seem to be unrelated to Aggregation.
They just so happen to be used in/from an aggregation plugin.

Maybe not a great name, but `NumericUtils`, `NumberUtils`, or `SchemaUtils` seems more appropriate.
What do you think?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java [hunk 14]
```diff
@@ -156,7 +156,7 @@
   }
 
   /**
-   * Common properties for BatchWritable source and sinks
+   * AggregationUtils properties for BatchWritable source and sinks
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java]: this javadoc seems wrong/confusing.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggregationUtils.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.aggregator;
+
+import co.cask.cdap.api.data.schema.Schema;
+
+/**
+ * Common functions for aggregation related functionalities.
```
Comment (anwar6953) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/AggregationUtils.java:22]: The function(s) defined in this utility class seem to be unrelated to Aggregation.
They just so happen to be used in/from an aggregation plugin.

Maybe not a great name, but `NumericUtils`, `NumberUtils`, or `SchemaUtils` seems more appropriate.
What do you think?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.aggregator.function;
+
+public final class RunningStats  {
+  private long entries = 0l;
+  private double oldM = 0d;
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: These names are very hard to read. Expand the abbreviation?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+  private double newM = 0d;
+  private double oldS = 0d;
+  private double newS = 0d;
+
+  public void push(double x) {
+    entries = entries + 1;
+    // See Knuth TAOCP vol 2, 3rd edition, page 232
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: Might be helpful to point to an online source too. e.g. https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance  (Online algorithm)

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+  public long values() {
+    return entries;
+  }
+
+  public double mean() {
+    return (entries > 0) ? newM : 0.0;
+  }
+
+  public double variance() {
+    return ( (entries > 1) ? newS /(entries - 1) : 0.0 );
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: Space after /

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  public double stddev() {
+    return Math.sqrt(variance());
```
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: Maybe add a note saying variance() can never return a negative value.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java (hunk 9 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java:24]: done!

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java (hunk 10 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: removed method. Not needed.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Variance.java (hunk 11 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/Variance.java:50]: Fixed the algo and now it doesn't matter.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java (hunk 14 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/common/Properties.java]: fixed

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java (hunk 16 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: If I knew what they were, I would have. Could you help with this ?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java (hunk 17 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: done!

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java (hunk 18 above)
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: new algo, so this is gone.

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java (hunk 10 above)
Comment (Nishith) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/function/RunningStats.java]: M -> mean
S -> stddev

**Comment by anwar6953** [comment]
Code changes look good.
There don't seem to be unit or integration test cases for this though.

**Comment by nitinmotgi** [comment]
@anwar6953 in the next PR.
